File: ruby/153_find_minimum_in_rotated_sorted_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_nums_head(nums)
    for i in 1..nums.size-1
        return i if nums[i] < nums[i-1]
    end
    0
end

# 找到被旋转的数组中最小的，其实也就是找到数组的head
# 因为数组是升序的，只要nums[i] < nums[i-1]，那么i就是head，返回[head] 即可

def find_min(nums)
    head = find_nums_head(nums)
    nums[head]
end

# 我们可以利用上述思路，从left，right两个方向判断，加快速度

def find_min_version2(nums)
	return -1 if nums.empty?
	return nums[0] if nums.size == 1 || nums[0] < nums[nums.size-1]

	left, right = 0, nums.size-1
	while left <= right
		return nums[left+1] if nums[left+1] < nums[left]
		return nums[right] if nums[right-1] > nums[right]
		left += 1
		right -= 1
	end

	-1
end

File: cpp/153_find_minimum_in_rotated_sorted_array.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 给定被旋转的一个有序数组，也就是数组中存在两段有序，让我们找到最小值；PS：数组中的元素没有重复；
// 使用二分，每次找到mid后判断左右两边是否存在不是有序，如果存在，则去不是有序的一边，最小的值一定是在非有序这边。
// 直到只要两个元素，此时左右都有序，返回其中较小的即可。
// 如果nums只要一个元素，那么不会进行二分，直接返回nums[left]或者nums[right]
class Solution {
public:
    int findMin(vector<int>& nums) {
		if (nums.empty()) { return -1; }
		int left = 0, right = nums.size() - 1;
		while (left < right) {
			int mid = left + (right - left) / 2;
			if (nums[left] > nums[mid]) {
				right = mid;
			} else if (nums[mid+1] > nums[right]) {
				left = mid+1;
			} else {
				return min(nums[left], nums[mid+1]);
			}
		}
		return nums[right];
    }
};
